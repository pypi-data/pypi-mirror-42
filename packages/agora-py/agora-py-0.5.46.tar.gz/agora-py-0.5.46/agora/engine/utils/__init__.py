"""
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Ontology Engineering Group
        http://www.oeg-upm.net/
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Copyright (C) 2016 Ontology Engineering Group.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

            http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
"""
import os
from threading import Lock

__author__ = 'Fernando Serena'


class Singleton(type):
    _lock = Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            id = kwargs.get('id', '')
            if id not in cls.instances:
                cls.instances[id] = super(Singleton, cls).__call__(*args, **kwargs)
            return cls.instances[id]


def prepare_store_path(base, path):
    if base and not os.path.exists(base):
        os.makedirs(base)
    if base or path:
        full_path = '/'.join(filter(lambda x: x, [base, path]))
        if not os.path.exists(full_path):
            os.makedirs(full_path)
        return full_path


class Semaphore(object):
    def __init__(self):
        self.value = -1

    def __enter__(self):
        self.value = 0

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.value = 1

    def isSet(self):
        return self.value == 1

    def set(self):
        self.value = 1


def get_immediate_subdirectories(a_dir):
    if not os.path.exists(a_dir):
        os.makedirs(a_dir)
    return [name for name in os.listdir(a_dir)
            if os.path.isdir(os.path.join(a_dir, name))]


class Wrapper(object):
    def __init__(self, object):
        self.__lock = Lock()
        self.__attr_locks = {}

        if isinstance(object, Wrapper):
            self.__object = object.__object
            self.__argret = object.__argret
            self.__attr_locks = object.__attr_locks
            self.__lock = object.__lock
        else:
            self.__object = object
            self.__argret = {}

    def decorate(self, f):
        def wrapper(*args, **kwargs):

            key = f.func_name + str(args)
            with self.__lock:
                if key not in self.__attr_locks:
                    self.__attr_locks[key] = Lock()

            with self.__attr_locks[key]:
                if key not in self.__argret:
                    ret = f(*args, **kwargs)

                    if kwargs:
                        return ret
                    if ret is not None:
                        self.__argret[key] = ret

                return self.__argret.get(key, None)

        return wrapper

    def __getattr__(self, item):

        if item.startswith('__'):
            return self.__object.__getattribute__(item)

        with self.__lock:
            if item not in self.__attr_locks:
                self.__attr_locks[item] = Lock()

        with self.__attr_locks[item]:
            if item not in self.__argret:
                attr = self.__object.__getattribute__(item)
                if hasattr(attr, '__call__'):
                    return self.decorate(attr)
                else:
                    self.__argret[item] = attr

        return self.__argret[item]


stopped = Semaphore()
