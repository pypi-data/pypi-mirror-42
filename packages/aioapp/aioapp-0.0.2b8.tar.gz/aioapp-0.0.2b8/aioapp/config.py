import copy
import os
from collections import OrderedDict
from typing import Optional, Union, Dict, Any, Type
from os import _Environ

Env = Union[_Environ, dict]


class ConfigError(Exception):
    pass


class Val:

    def __init__(self, name: str, value: Any) -> None:
        self.name = name
        self.value = value

    def __call__(self):  # pragma: nocover
        raise NotImplementedError()

    @staticmethod
    def type_name() -> str:
        return ''

    def args_markdown(self) -> str:
        return ''


class StrVal(Val):

    def __init__(self, name: str, value: Any,
                 max: Optional[int] = None,
                 min: Optional[int] = None) -> None:
        super().__init__(name, value)
        self.min = min
        self.max = max

    def __call__(self) -> str:
        if not isinstance(self.value, str):
            raise ConfigError("%s must be a string" % self.name)
        if self.min is not None and len(self.value) < self.min:
            raise ConfigError("length of %s must be greater than or equal to "
                              "%s"
                              "" % (self.name, self.min))
        if self.max is not None and len(self.value) > self.max:
            raise ConfigError("length of %s must be less than or equal to %s"
                              "" % (self.name, self.max))

        return self.value

    @staticmethod
    def type_name() -> str:
        return 'string'

    def args_markdown(self) -> str:
        text = ''
        if self.min is not None:
            text += '\n  min length: %s' % self.min
        if self.max is not None:
            text += '\n  max length: %s' % self.max
        return text


class BoolVal(Val):

    def __call__(self) -> bool:
        if isinstance(self.value, bool):
            return self.value
        if isinstance(self.value, int):
            return self.value != 0
        if isinstance(self.value, str):
            if self.value.lower() in ('1', 'on', 'true', 't'):
                return True
            if self.value.lower() in ('0', 'off', 'false', 'f'):
                return False
        raise ConfigError("%s must be a boolean" % self.name)

    @staticmethod
    def type_name() -> str:
        return 'boolean'


class IntVal(Val):

    def __init__(self, name: str, value: Any,
                 max: Optional[int] = None,
                 min: Optional[int] = None) -> None:
        super().__init__(name, value)
        self.min = min
        self.max = max

    def __call__(self) -> int:
        try:
            val = int(self.value)
        except Exception:
            raise ConfigError("%s must be an integer" % self.name)
        if self.min is not None and val < self.min:
            raise ConfigError("%s must be greater than or equal to %s"
                              "" % (self.name, self.min))
        if self.max is not None and val > self.max:
            raise ConfigError("%s must be less than or equal to %s"
                              "" % (self.name, self.max))
        return val

    @staticmethod
    def type_name() -> str:
        return 'integer'

    def args_markdown(self) -> str:
        text = ''
        if self.min is not None:
            text += '\n  min value: %s' % self.min
        if self.max is not None:
            text += '\n  max value: %s' % self.max
        return text


class FloatVal(Val):

    def __init__(self, name: str, value: Any,
                 max: Optional[float] = None,
                 min: Optional[float] = None) -> None:
        super().__init__(name, value)
        self.min = min
        self.max = max

    def __call__(self) -> float:
        try:
            val = float(self.value)
        except Exception:
            raise ConfigError("%s must be a float" % self.name)
        if self.min is not None and val < self.min:
            raise ConfigError("%s must be greater than or equal to %s"
                              "" % (self.name, self.min))
        if self.max is not None and val > self.max:
            raise ConfigError("%s must be less than or equal to %s"
                              "" % (self.name, self.max))
        return val

    @staticmethod
    def type_name() -> str:
        return 'float'

    def args_markdown(self) -> str:
        text = ''
        if self.min is not None:
            text += '\n  min value: %s' % self.min
        if self.max is not None:
            text += '\n  max value: %s' % self.max
        return text


class FileVal(Val):

    def __init__(self, name: str, value: Any,
                 mode: str = 'r',
                 encoding: str = 'UTF-8') -> None:
        super().__init__(name, value)
        self.mode = mode
        self.encoding = encoding

    def __call__(self, mode: str = 'r', encoding: str = 'UTF-8') -> str:
        try:
            with open(self.value, mode, encoding=encoding):
                pass
        except Exception as e:
            raise ConfigError("Could not access to file %s with error: %s"
                              "" % (self.value, e))
        return self.value

    @staticmethod
    def type_name() -> str:
        return 'string(path to file)'

    def args_markdown(self) -> str:
        text = ''
        if self.mode is not None:
            text += '\n  assess mode: %s' % self.mode
        if self.encoding is not None:
            text += '\n  encoding: %s' % self.encoding
        return text


class DirVal(Val):

    def __call__(self) -> str:
        if not os.path.exists(self.value):
            raise ConfigError("Directory %s does not exist"
                              "" % self.value)
        return self.value

    @staticmethod
    def type_name() -> str:
        return 'string(path to dir)'


class Config:
    _vars: Dict[str, Union[Dict, OrderedDict]] = {}

    def __init__(self,
                 env: Optional[Env] = None) -> None:
        self._env = env or os.environ
        self._conf = copy.deepcopy(self._vars)
        self._description: Dict[str, Dict] = {}
        for key, val in self._conf.items():
            val_type = val.pop('type')
            val_name = val.pop('name')
            val_default = None
            if 'default' in val:
                val_default = val.pop('default')
            required = False
            if 'required' in val:
                required = bool(val.pop('required'))
            descr = ''
            if 'descr' in val:
                descr = str(val.pop('descr'))

            self._description[val_name] = {
                'type': val_type,
                'default': val_default,
                'required': required,
                'descr': descr,
            }

            value = self._env.get(val_name, val_default)
            if value is None:
                if required is True:
                    raise ConfigError("%s is required" % val_name)
                else:
                    setattr(self, key, None)
            else:
                v: Any = self._get_val(val_type)
                setattr(self, key, v(val_name, value, **val)())

    @classmethod
    def _get_val(cls, val_type: Any) -> Type[Val]:
        if val_type == str:
            return StrVal
        elif val_type == bool:
            return BoolVal
        elif val_type == int:
            return IntVal
        elif val_type == float:
            return FloatVal
        elif val_type == 'file':
            return FileVal
        elif val_type == 'dir':
            return DirVal
        elif isinstance(val_type, type) and issubclass(val_type, Val):
            return val_type
        else:
            raise UserWarning('Invalid configuration settings')

    @classmethod
    def as_markdown(cls):
        result = []

        for name, options in cls._vars.items():
            options = copy.deepcopy(options)

            type_cls = cls._get_val(options.pop('type'))
            env_name = options.pop('name')
            descr = ''
            if 'descr' in options:
                descr = ': %s' % options.pop('descr')
            text = '* %s%s\n  type: %s' % (env_name, descr,
                                           type_cls.type_name())
            if 'required' in options:
                required = options.pop('required')
                if required:
                    text += '\n\n  required'
            if 'default' in options:
                default = options.pop('default')
                if default:
                    text += '\n  default: %s' % default

            inst = type_cls(env_name, None, **options)
            text += inst.args_markdown()

            result.append(text)

        return '\n\n'.join(result) + '\n'
