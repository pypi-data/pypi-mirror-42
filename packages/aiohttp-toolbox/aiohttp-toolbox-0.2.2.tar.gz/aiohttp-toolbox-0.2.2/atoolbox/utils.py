import json
import re
from typing import Any, Tuple, Type, TypeVar, Union

from aiohttp.web import Response
from aiohttp.web_exceptions import HTTPException
from pydantic import BaseModel, ValidationError, validate_model
from pydantic.fields import Shape

from .json_tools import pretty_lenient_json

JSON_CONTENT_TYPE = 'application/json'
IP_HEADER = 'X-Forwarded-For'
PROTO_HEADER = 'X-Forwarded-Proto'
URI_NOT_ALLOWED = re.compile(r'[^a-zA-Z0-9_\-/.]')
REMOVE_PORT = re.compile(r':\d{2,}$')
PydanticModel = TypeVar('PydanticModel', bound=BaseModel)

__all__ = (
    'raw_json_response',
    'json_response',
    'parse_request_json',
    'parse_request_json_ignore_missing',
    'parse_request_query',
    'get_ip',
    'request_root',
    'JsonErrors',
    'encrypt_json',
    'decrypt_json',
    'get_offset',
    'slugify',
    'remove_port',
    'RequestError',
)


def raw_json_response(json_str: Union[str, bytes, None], status_: int = 200):
    if isinstance(json_str, str):
        body = json_str.encode()
    elif isinstance(json_str, bytes):
        body = json_str
    elif json_str is None:
        body = b'null'
    else:
        raise TypeError(f'json_str must be bytes or str, not "{type(json_str)}')
    return Response(body=body + b'\n', status=status_, content_type=JSON_CONTENT_TYPE)


def json_response(*, status_=200, list_=None, headers_=None, **data):
    return Response(
        body=json.dumps(data if list_ is None else list_).encode() + b'\n',
        status=status_,
        content_type=JSON_CONTENT_TYPE,
        headers=headers_,
    )


async def parse_request_json(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    error_details = None
    try:
        data = await request.json()
    except ValueError:
        error_msg = 'Invalid JSON'
    else:
        try:
            return model.parse_obj(data)
        except ValidationError as e:
            error_msg = 'Invalid Data'
            error_details = e.errors()

    raise JsonErrors.HTTPBadRequest(message=error_msg, details=error_details, headers=headers)


def parse_request_query(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    data = {}
    for k in request.query:
        v = request.query.getall(k)
        f = model.__fields__.get(k)
        if len(v) > 1 or f and f.shape != Shape.SINGLETON:
            data[k] = v
        else:
            data[k] = v[0]

    try:
        return model(**data)
    except ValidationError as e:
        raise JsonErrors.HTTPBadRequest(message='Invalid query data', details=e.errors(), headers=headers)


async def parse_request_json_ignore_missing(
    request, model: Type[PydanticModel], *, headers=None
) -> Tuple[PydanticModel, dict]:
    try:
        raw_data = await request.json()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(message='Invalid JSON', headers=headers)
    if not isinstance(raw_data, dict):
        raise JsonErrors.HTTPBadRequest(message='data not a dictionary', headers=headers)

    data, e = validate_model(model, raw_data, raise_exc=False)
    if e:
        errors = [e for e in e.errors() if not (e['type'] == 'value_error.missing' and len(e['loc']) == 1)]
        if errors:
            raise JsonErrors.HTTPBadRequest(message='Invalid Data', details=errors, headers=headers)

    return model.construct(**data), raw_data


def get_ip(request):
    ips = request.headers.get(IP_HEADER)
    if ips:
        return ips.split(',', 1)[0].strip(' ')
    else:
        return request.remote


def request_root(request):
    # request.url.scheme doesn't work as https is already terminated
    scheme = request.headers.get(PROTO_HEADER) or 'http'
    return f'{scheme}://{request.host}'


class JsonErrors:
    class _HTTPExceptionJson(HTTPException):
        custom_reason = None

        def __init__(self, message, *, details=None, headers=None):
            self.message = message
            self.details = details
            data = {'message': message}
            if details:
                data['details'] = details
            super().__init__(
                text=pretty_lenient_json(data),
                content_type=JSON_CONTENT_TYPE,
                headers=headers,
                reason=self.custom_reason,
            )

        def __repr__(self) -> str:
            return f'{super().__repr__()}, {self.status}: {self.message}'

        def __str__(self) -> str:
            return repr(self)

    class HTTPAccepted(_HTTPExceptionJson):
        status_code = 202

    class HTTPBadRequest(_HTTPExceptionJson):
        status_code = 400

    class HTTPUnauthorized(_HTTPExceptionJson):
        status_code = 401

    class HTTPPaymentRequired(_HTTPExceptionJson):
        status_code = 402

    class HTTPForbidden(_HTTPExceptionJson):
        status_code = 403

    class HTTPNotFound(_HTTPExceptionJson):
        status_code = 404

    class HTTPMethodNotAllowed(_HTTPExceptionJson):
        status_code = 405

        def __init__(self, message, allowed_methods, *, headers=None):
            headers = headers or {}
            headers.setdefault('Allow', ','.join(allowed_methods))
            super().__init__(message, details={'allowed_methods': allowed_methods}, headers=headers)

    class HTTPConflict(_HTTPExceptionJson):
        status_code = 409

    class HTTP470(_HTTPExceptionJson):
        status_code = 470
        custom_reason = 'Invalid user input'


def encrypt_json(app, data: Any) -> str:
    return app['auth_fernet'].encrypt(json.dumps(data).encode()).decode()


def decrypt_json(app, token: bytes, *, ttl: int = None, headers=None) -> Any:
    from cryptography.fernet import InvalidToken

    try:
        return json.loads(app['auth_fernet'].decrypt(token, ttl=ttl).decode())
    except InvalidToken:
        raise JsonErrors.HTTPBadRequest(message='invalid token', headers=headers)


def get_offset(request, paginate_by=100):
    page = request.query.get('page')
    if not page:
        return 0

    try:
        p = int(page)
        if p < 1:
            raise ValueError()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(f"invalid page '{page}'")
    else:
        return (p - 1) * paginate_by


def slugify(title):
    name = title.replace(' ', '-').lower()
    name = URI_NOT_ALLOWED.sub('', name)
    name = re.sub('-{2,}', '-', name)
    return name.strip('_-')


def remove_port(host):
    return REMOVE_PORT.sub('', host)


class RequestError(RuntimeError):
    def __init__(self, status, url, *, text: str = None):
        self.status = status
        self.url = url
        self.text = text

    def __str__(self):
        return f'response {self.status} from "{self.url}"' + (f':\n{self.text[:400]}' if self.text else '')

    def json(self):
        return json.loads(self.text)

    def extra(self):
        return self.text
