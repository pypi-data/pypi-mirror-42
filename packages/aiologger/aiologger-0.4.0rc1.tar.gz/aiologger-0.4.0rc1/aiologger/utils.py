from typing import Callable


class classproperty:
    def __init__(self, func):
        self._func = func

    def __get__(self, obj, owner):
        return self._func(owner)


class CallableWrapper:
    def __init__(self, func: Callable) -> None:
        self.func = func

    def __call__(self, *args, **kwargs):
        return self.func(*args, **kwargs)
