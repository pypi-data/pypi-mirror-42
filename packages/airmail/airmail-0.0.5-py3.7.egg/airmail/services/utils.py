import os, sys
from pathlib import Path
import click
import yaml

def err_log(msg, *args):
    if args:
        msg %= args
    click.echo(click.style('Error: ', fg='red', bold=True) + msg)

# Read a yaml file given a file path. Shou
def read_yml(filepath):
    target_file = Path(filepath)

    if target_file.is_file():
        with open(target_file) as yml_file:
            return yaml.load(yml_file)
    else:
        click.echo(click.style('Error: ', fg='red', bold=True) +
            'Could not read the .yml at ' +
            click.style(filepath, bold=True))
        sys.exit()

# Determines the path to files inside the
# project. (i.e. the `data` directory)
def determine_project_path():
    """Borrowed from wxglade.py"""
    try:
        root = __file__
        if os.path.islink (root):
            root = os.path.realpath (root)
        return os.path.dirname (os.path.abspath (root))
    except:
        print("I'm sorry, but something is wrong.")
        print("There is no __file__ variable. Please contact the author.")
        sys.exit ()
