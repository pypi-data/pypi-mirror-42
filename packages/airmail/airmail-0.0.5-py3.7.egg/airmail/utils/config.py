from functools import reduce

def build_config(transforms, config):
    """Reduces a dictionary through functions to modify the dictionary"""
    return reduce(lambda c, fn: fn(c), transforms, config)
