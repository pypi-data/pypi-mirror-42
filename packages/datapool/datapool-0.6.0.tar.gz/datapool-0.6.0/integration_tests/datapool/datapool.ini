[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

