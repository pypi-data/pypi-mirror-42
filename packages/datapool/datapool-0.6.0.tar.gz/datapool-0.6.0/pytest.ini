[pytest]
addopts = -rs --full-trace
testpaths = tests
