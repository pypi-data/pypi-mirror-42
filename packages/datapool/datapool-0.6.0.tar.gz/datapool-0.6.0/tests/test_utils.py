# encoding: utf-8
from __future__ import absolute_import, division, print_function

import os

import pytest

from datapool.config import MagicConfig
from datapool.utils import (
    enumerate_filename,
    is_server_running,
    list_folder_recursive,
    parse_timestamp,
    remove_pid_file,
    warn_if_generator_is_not_used,
    write_pid_file,
)


def test_folder_list(data_path, regtest):
    for path in list_folder_recursive(data_path("../etc"), also_folders=False):
        print(path, file=regtest)

    for path in list_folder_recursive(data_path("../etc"), also_folders=True):
        print(path, file=regtest)


def test_enumerate_filename(tmpdir):
    fldr = tmpdir.strpath
    fname = tmpdir.join("abc").strpath

    def ef(p):
        return enumerate_filename(p)[0]

    assert ef(fname) == os.path.join(fldr, "abc_0")
    assert ef(fname) == os.path.join(fldr, "abc_0")

    fname = tmpdir.join("abc.txt").strpath
    assert ef(fname) == os.path.join(fldr, "abc_0.txt")
    next_name = ef(fname)
    assert next_name == os.path.join(fldr, "abc_0.txt")

    def touch(path):
        open(path, "w").close()
        return path

    touch(fname)
    touch(next_name)
    next_name = ef(fname)
    assert next_name == os.path.join(fldr, "abc_1.txt")

    touch(next_name)
    next_name = ef(fname)
    assert next_name == os.path.join(fldr, "abc_2.txt")

    touch(os.path.join(fldr, "abc_9.txt"))
    next_name = ef(fname)
    assert next_name == os.path.join(fldr, "abc_10.txt")

    fname_1 = touch(os.path.join(fldr, "xyz.txt"))
    fname_2 = touch(os.path.join(fldr, "uvw_3.txt"))

    next_1, next_2 = enumerate_filename(fname_1, fname_2)
    assert next_1 == os.path.join(fldr, "xyz_4.txt")
    assert next_2 == os.path.join(fldr, "uvw_4.txt")


def test_pid_functions(monkeypatch, tmpdir, regtest):

    config = MagicConfig()
    config.server.pid_file = tmpdir.join("datapool.pid").strpath
    config.print_(fh=regtest)

    assert is_server_running(config) is False
    write_pid_file(config)
    assert is_server_running(config) is True
    remove_pid_file(config)
    assert is_server_running(config) is False


def test_parse_timestamp(regtest):
    p = "2017 2017-10 2017-10-01 2017-10-01:23 2017-10-01:23:12 2017-10-01:23:12:01"
    for pi in p.split(" "):
        print(str(parse_timestamp(pi)), file=regtest)

    p = "17 2017-13 2017-10-32 2017-10-01:25 2017-10-01:23:77 2017-10-01:23:12:71"
    for pi in p.split(" "):
        with pytest.raises(ValueError):
            parse_timestamp(pi)


def test_warn_if_generator_is_not_used():
    @warn_if_generator_is_not_used
    def one_element():
        yield 1

    i = one_element()
    assert not i.iterated
    list(i)
    assert i.iterated

    @warn_if_generator_is_not_used
    def empty():
        return
        yield

    i = empty()
    assert not i.iterated
    list(i)
    assert i.iterated

    with pytest.raises(AssertionError):

        @warn_if_generator_is_not_used
        def classic_function():
            return
