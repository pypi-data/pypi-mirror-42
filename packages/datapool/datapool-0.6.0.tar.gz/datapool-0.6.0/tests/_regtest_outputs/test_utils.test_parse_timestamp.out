2017-01-01 00:00:00
2017-10-01 00:00:00
2017-10-01 00:00:00
2017-10-01 23:00:00
2017-10-01 23:12:00
2017-10-01 23:12:01
